### ISSUE TITLE
[new sysrepo] Internal error on update event

### ISSUE THREAD

**Issue by gitter-runner** [sysrepo/sysrepo#1765]
Hi,

I have two applications subscribed for the same 'config true' list with _SR_SUBSCR_UPDATE_.
On a PowerPC based system when I try to change data in the list I often get a timeout with the following error print:

`[ERR]: Internal error (/opt/dev/sysrepo/src/shm_sub.c:1901).`

This occurs quite often on this platform (approx. 5/3).

I've found out that it occurs when the internal update event in an application is handled before the other app starts processing it.
In this case _lyd_parse_mem()_ function (at shm_sub.c:1900) tries to parse an empty LYB data structure in the second app.

I attached a modified example application which can be used to reproduce it on x86 as well (using event processing delay):

```
# subscribe ietf-system /ietf-system:system/hostname &
# subscribe ietf-system /ietf-system:system/hostname delay &
# sr_set_item_example /ietf-system:system/hostname testhost

Application will watch for changes in "/ietf-system:system/hostname".


 ========== LISTENING FOR CHANGES ==========


Application will watch for changes in "/ietf-system:system/hostname".


 ========== LISTENING FOR CHANGES ==========


Application will set "/ietf-system:system/hostname" to "testhost".


 ========== EVENT update ====================================

[ERR]: Internal error (/opt/dev/sysrepo/src/shm_sub.c:1901).
[ERR]: Callback event "update" with ID 1 processing timed out.
[WRN]: Event "update" with ID 1 priority 0 failed (Timeout expired).
```

[subscribe_c.txt](https://github.com/sysrepo/sysrepo/files/4077512/subscribe_c.txt)

sysrepo version: 1.3.48
libyang version: 1.0.119

Thanks!

**Comment by michalvasko** [comment]
Hi,
actually, I seem to remember this feature was designed in the way that there can only be one `SR_SUBSCR_UPDATE` subscriber because how would you know which callback new values should be used otherwise? But, obviously, I have not documented this limitation.

So, your use-case cannot work and your observation is correct, it is the result of using one memory offset for both input and output of this callback. Looking further at the code, it seems it would not be much work to support such a use-case. But you would first have to justify it well because this event alone adds quite some complexity into the whole mechanism and I would prefer not to allow even more complicated usage.

TLDR; why can you possibly want to have 2 UPDATE subscriptions on one list? They are just supposed to update the value in some rare circumstances and do nothing else, it does not make sense to have 2 callbacks for one value.

Regards,
Michal

**Comment by gitter-runner** [comment]
Thanks for your answer. I need to implement an apply callback with option to deny the changes and rollback the transaction.
I use _SR_EV_CHANGE_ event for it since _SR_EV_DONE_ cannot deny the changes.

Furthermore I need a verify callback as well. I used _SR_EV_UPDATE_ event for it.
So I don't want to change the values, only want to be notified about the changes before the _SR_EV_CHANGE_ phase.

Can you suggest a better way to implement similar behaviour without using _SR_SUBSCR_UPDATE_?

Thanks!

**Comment by michalvasko** [comment]
Hi,
why cannot you both verify the value and apply it in `SR_EV_CHANGE`?

Regards,
Michal

**Comment by gitter-runner** [comment]
Hi, 
I would like to optimize the number of config rollbacks by executing all validations before any change is sent to hardware. For example when there is a transaction affecting a lot of change callbacks, and the last fails on validation all the previous changes must be rolled back in the hardware which can take a lot of time.

**Comment by michalvasko** [comment]
Hi,
then you should make sure all these changes are applied in one callback or in as few as possible. First verify all the changes for the particular callback and then apply them. But this should not be a situation that happens often (that a callback refuses some changes) so we thought it acceptable even if the rollback takes a lot of time.

Regards,
Michal
